### PR TITLE
{tools}[GCCcore/12.2.0] IPython v8.14.0, JupyterHub v4.0.1, JupyterLab v4.0.3, ...

### PR DESCRIPTION
**Pull request by arielzn**
(created using `eb --new-pr`)

### CODE REVIEW

**Comment by arielzn** [comment]
these requires #18404 #18405  and #18406

**Comment by arielzn** [comment]
On this PR apart of adding newer versions on 2022b, there's some cleanup/reorganization done on the extensions list for each bundle. 

Basically for IPython we were including before several extensions that were exclusively  jupyter requirements (jupyter_core, jupyter_client, nb* ).
By using the helper script  `findPythonDeps.py` these are now moved into either the jupyter-server or JupyterLab ebs accordingly. 

I've took note of the fix #17612 and added the correct versions for them but on the JupyterLab eb, as they are meant to be used with it not ipython.

**Comment by casparvl** [comment]
@arielzn while I am all for cleaning up, I am missing one crucial thing in the list now: the classical notebook interface. This used to be part of the `IPython` module, by means of the `notebook` package. That was _also_ the reason that things like `jupyter_core`, `jupyter_client`, `widgetsnbextension` etc were part of the IPython EasyConfig.

It makes me wonder what is the right place for what. I think we have two options:
1. Essentially, keep things as they were before. That is: `IPython` supplies the classic Notebook server functionality. That means we'll have to retain things like the list I described above in this EasyConfig. Since `widgetsnbextension` is also relevant for the classic Notebook server, it should _probably_ stay in there (technically it is an optional dependency, but I think most sites will appreciate having this supported/installed by default). Note that this means our `IPython` EasyConfig is essentially what IPython _used_ to be when it was still a monolothic release (up to 3.X, see "Jupyter and the future of IPython" on https://ipython.org/).

2. Split things up, like you propose. However, I'd then say we also need a `jupyter-notebook` EasyConfig, to replace the `notebook` package that has been taken out from `IPython`. One thing to note is that while I recently pushed `nbclassic` into `jupyter-server` ([here](https://github.com/easybuilders/easybuild-easyconfigs/pull/18389)), one could argue that in this new setup, it would fit better in `jupyter-notebook` - I'm not sure. In my view, the widgets functionality should then be put in a standalone EasyConfig as well, so that it can be used as a depencency both in `jupyter-notebook` and `JupyterLab`.

One could argue that 2 fits better with what IPython devs themselves did, which is split this up. The IPython module then 'only' provides the Interactive Python, part of which is to provide a Python kernel to Jupyter. It would mean that in practice, I would end up with quite a list of package to load for the things I'd like to offer to my users. Probably, I would load:
```
IPython
jupyter-notebook
jupyter-server
jupyter-widgets (or whatever we'd call this standalone thing)
JupyterLab
JupyterHub
batschpawner
jupyterlmod
jupyter-resource-usage
```
Not a huge issue, I could always create a Bundle out of it if it became too annoying. 

Overall, I don't think I have a very strong preference. Maybe I'm leaning a bit towards solution (2), as it is more modular: some sites might want to offer (only) a JupyterLab, or classic Jupyter Notebook interface, others might like to offer both (even if offering both hurts no one...). For `jupyter-widgets`, we can argue if it should be a dependency for `JupyterLab` and/or `jupyter-notebook` (though I'd prefer yes, since it is _such_ a commonly used extension, people essentially expect it to be there), but in any case it should be stand alone so that the same dependency can be reused between `JupyterLab` and `jupyter-notebook` (this also guarantees you can load _both_ `JupyterLab` and `jupyter-notebook` and offer both simultanously).

Just as a side note: we now offer a JupyterHub (based on 2022a toolchain) where JupyterLab is loaded by default, but for reasons of continuity, we still offer the notebook interface by default by setting `export JUPYTERHUB_SINGLEUSER_APP='notebook.notebookapp.NotebookApp'` ([see here](https://jupyterhub.readthedocs.io/en/stable/howto/configuration/config-user-env.html#switching-back-to-the-classic-notebook)). Just to say that having both the classic and jupyterlab interface available makes a lot of sense and is something we should make sure works - whatever route we choose. There is even a button (help => classic notebook interface, or something like that) in the JupyterLab interface to switch _back_ to the notebook interface.

@lexming maybe you also have an opinion on this, since you fixed the previous widget-version related issue?

**Comment by arielzn** [comment]
@casparvl  earlier today I saw your fix  #18389  and started looking into this actually. 

What I did is this latest push, where I'm adding nbclassic into jupyter-server as you did. 

Regarding your proposals, IMHO the IPython eb was badly needing a cleanup, I read before starting all this new easyconfigs that section you mention from ipython.org, notably: `As of IPython 4.0, the language-agnostic parts of the project: the notebook format, message protocol, qtconsole, notebook web application, etc. have moved to new projects under the name [Jupyter](https://jupyter.org/).` 

So having on the IPython bundle all `jupyter_*` and `nb*` extensions included, is going against the devs decision made 4 major releases ago for it.

About where to locate the classic notebook stuff after seeing your fix for the 2022a version, I've did the same and added nbclassic into jupyter-server, I've just pushed this modification.

For the widgets I've located them previously on the JupyterLab eb, but is true that they can't be used with the classic notebook provided by juypter-server as in my latest push.

So we could also move the 3 widgets extensions to the jupyter-server eb, thus having the same split as in 2022a:

IPython
jupyter-server   (here I added nbclassic and we could also move the 3 widgets ones ?)
JupyterLab
JupyterHub    (we keep batschpawner  inside here also ?)
jupyterlmod
jupyter-resource-usage


I remark that on all these easyconfigs bundles, I'm explicitly ordering the extension list naming **after** the extension corresponding to the eb name (jupyter-server, JupyterLab, JupyterHub)  extra stuff which is bundled together for convenience.

On JupyterHub we are having apart of batchspwaner, several others related with authentication  stuff (juypterhub-***authenticator), I think it makes sense to keep them together for those rather than going for all strict split.

**Comment by arielzn** [comment]
I've just pushed a commit according to my latest comments. Making `jupyter-server` fully functional as classic notebook server with widgets.

I guess is easier to test this possibility with the commit providing that already there.

**Comment by casparvl** [comment]
Thanks for the quick changes. I'll test them in our setup, but that'll take me a bit of time, so don't know if I will get to do that today.

On thing I'm still confused about: you seem to make the assumption that `nbclassic` is all you need to provide the classic notebook interface. I actually thought one needed `notebook` and `nbclassic` together - but that assumption was probably completely wrong. What confused me a bit is what the rolw of `notebook` then is today. From the `nbclassic` docs:

> Because NbClassic provides the classic interface on top of the new [Jupyter Server](https://github.com/jupyter/jupyter_server/) backend, it can coexist with other frontends like JupyterLab and Notebook 7 in the same installation. NbClassic preserves the custom classic notebook experience under a new set of URL endpoints, under the namespace /nbclassic/.

I'm guessing you are right: you probably _don't_ need `notebook` to serve a classic notebook interface if you use `jupyter-server` + `nbclassic`. I guess the `notebook` package is then only relevant if you want to run a Jupyter Notebook Server _without_ running on top of `jupyter-server`? In other words, I guess I have two options of offering a classic notebook interface:

`jupyter-server` + `nbclassic`

or

`notebook`

If, like me, you want to offer anything else that uses `jupyter-server` as a dependency (e.g. `JupyterLab` or `jupyter-server-proxy`), I guess I _have_ to go for the first option.

It also means there is still a valid use case for creating a `jupter-notebook` EasyConfig, which contains `notebook`, correct? (not saying that has to be part of this PR - I don't mind if it isn't. To me, the current combination in your PR provides all the functionality I'd need :))

**Comment by casparvl** [comment]
Ran into the issue that I previously fixed in https://github.com/easybuilders/easybuild-easyconfigs/pull/18390 so I updated batchspawner to a specific commit (support for JH 3.X and 4.X hasn't made it to an official release yet).

**Comment by casparvl** [comment]
So, I tested, and found a few things:

- With `module load JupyterHub/4.0.1-GCCcore-12.2.0` and `module load JupyterLab/4.0.3-GCCcore-12.2.0`, I can start a Jupyter-server instance that will serve me the lab environment. Using the 'help' function, I can switch back to the classic notebook interface. I do see
```
[I 2023-08-01 13:41:18.322 ServerApp] nbclassic | extension was successfully loaded.
[W 2023-08-01 13:41:18.322 ServerApp] notebook | extension failed loading with message: ExtensionLoadingError('_load_jupyter_server_extension function was not found.')
    Traceback (most recent call last):
      File "/scratch-nvme/1/casparl/generic/software/jupyter-server/2.7.0-GCCcore-12.2.0/lib/python3.10/site-packages/jupyter_server/extension/manager.py", line 356, in load_extension
        extension.load_all_points(self.serverapp)
      File "/scratch-nvme/1/casparl/generic/software/jupyter-server/2.7.0-GCCcore-12.2.0/lib/python3.10/site-packages/jupyter_server/extension/manager.py", line 228, in load_all_points
        return [self.load_point(point_name, serverapp) for point_name in self.extension_points]
      File "/scratch-nvme/1/casparl/generic/software/jupyter-server/2.7.0-GCCcore-12.2.0/lib/python3.10/site-packages/jupyter_server/extension/manager.py", line 228, in <listcomp>
        return [self.load_point(point_name, serverapp) for point_name in self.extension_points]
      File "/scratch-nvme/1/casparl/generic/software/jupyter-server/2.7.0-GCCcore-12.2.0/lib/python3.10/site-packages/jupyter_server/extension/manager.py", line 219, in load_point
        return point.load(serverapp)
      File "/scratch-nvme/1/casparl/generic/software/jupyter-server/2.7.0-GCCcore-12.2.0/lib/python3.10/site-packages/jupyter_server/extension/manager.py", line 146, in load
        loader = self._get_loader()
      File "/scratch-nvme/1/casparl/generic/software/jupyter-server/2.7.0-GCCcore-12.2.0/lib/python3.10/site-packages/jupyter_server/extension/manager.py", line 115, in _get_loader
        loader = get_loader(loc)
      File "/scratch-nvme/1/casparl/generic/software/jupyter-server/2.7.0-GCCcore-12.2.0/lib/python3.10/site-packages/jupyter_server/extension/utils.py", line 47, in get_loader
        raise ExtensionLoadingError(msg) from None
    jupyter_server.extension.utils.ExtensionLoadingError: _load_jupyter_server_extension function was not found.
```
In the jupyter-server output. That suggest to me that nbclassic and notebook are considered two different extensions, of which one is found, and the other is not. I wonder what having the `notebook` extension would do though, since the classic interface seems to be working just fine.. 

- With `module load JupyterHub/4.0.1-GCCcore-12.2.0` and `module load JupyterLab/4.0.3-GCCcore-12.2.0` _and_ `export JUPYTERHUB_SINGLEUSER_APP='notebook.notebookapp.NotebookApp'`, I get this error reported by the jupyter server:
```
Traceback (most recent call last):
  File "/scratch-nvme/1/casparl/generic/software/JupyterHub/4.0.1-GCCcore-12.2.0/bin/batchspawner-singleuser", line 8, in <module>
    sys.exit(main())
  File "/scratch-nvme/1/casparl/generic/software/JupyterHub/4.0.1-GCCcore-12.2.0/lib/python3.10/site-packages/batchspawner/singleuser.py", line 39, in main
    run_path(cmd_path, run_name="__main__")
  File "/scratch-nvme/1/casparl/generic/software/Python/3.10.8-GCCcore-12.2.0/lib/python3.10/runpy.py", line 289, in run_path
    return _run_module_code(code, init_globals, run_name,
  File "/scratch-nvme/1/casparl/generic/software/Python/3.10.8-GCCcore-12.2.0/lib/python3.10/runpy.py", line 96, in _run_module_code
    _run_code(code, mod_globals, init_globals,
  File "/scratch-nvme/1/casparl/generic/software/Python/3.10.8-GCCcore-12.2.0/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/scratch-nvme/1/casparl/generic/software/JupyterHub/4.0.1-GCCcore-12.2.0/bin/jupyterhub-singleuser", line 5, in <module>
    from jupyterhub.singleuser import main
  File "/scratch-nvme/1/casparl/generic/software/JupyterHub/4.0.1-GCCcore-12.2.0/lib/python3.10/site-packages/jupyterhub/singleuser/__init__.py", line 67, in <module>
    from .app import SingleUserNotebookApp, main
  File "/scratch-nvme/1/casparl/generic/software/JupyterHub/4.0.1-GCCcore-12.2.0/lib/python3.10/site-packages/jupyterhub/singleuser/app.py", line 31, in <module>
    App = import_item(JUPYTERHUB_SINGLEUSER_APP)
  File "/scratch-nvme/1/casparl/generic/software/IPython/8.14.0-GCCcore-12.2.0/lib/python3.10/site-packages/traitlets/utils/importstring.py", line 30, in import_item
    module = __import__(package, fromlist=[obj])
ModuleNotFoundError: No module named 'notebook'
```
This confuses me further: is the notebook package needed, or not? In the first scenario, it doesn't seem like it. In the second case, it does. Maybe because that environment variable references the `notebook` package specifically, and not `nbclassic`? It is however the official instruction from [here](https://jupyterhub.readthedocs.io/en/stable/howto/configuration/config-user-env.html#switching-back-to-the-classic-notebook).

Every time I think I understand the Jupyter ecosystem, I get confused again xD 

Just now, I found [this post](https://discourse.jupyter.org/t/difference-in-arguments-between-jupyter-notebook-and-jupyter-lab/11179) and [this post](https://stackoverflow.com/questions/67797152/what-is-the-difference-between-jupyter-notebook-and-jupyter-server/67804732#67804732) both of which were helpful to understand. I'll copy the key points here in case they ever disappear:

```
The main, hopefully invisible, difference is in the server which will be used:

- jupyter notebook will start the old notebook server
- jupyter lab or jupyter nbclassic will start the new jupyter_server server (“this is the future”)
...
There are no more differences for jupyter notebook and jupyter nbclassic (if anything else differs, it should probably be fixed). If you use JupyterLab 3.x, when starting jupyter lab and switching to classic notebook interface you were already using nbclassic, probably without even noticing!
```

Ok, I think I understand why that worked then, whereas with `export JUPYTERHUB_SINGLEUSER_APP='notebook.notebookapp.NotebookApp'` it does not. Switching to the classic interface via the `help` menu in JupyterLab actually makes you switch to `nbclassic` (i.e. the classic notebook interface running on top of jupyter-server). Using `export JUPYTERHUB_SINGLEUSER_APP='notebook.notebookapp.NotebookApp'`, you actually switch to the _traditional notebook server_ from the `notebook` package. That is not installed by this PR, hence the error. I think essentially that corresponds to what is written [here](https://jupyter-notebook.readthedocs.io/en/latest/migrating/multiple-interfaces.html#nbclassic-and-notebook-7)

From what I undestand from all of it, there is _no point_ in having `nbclassic` _and_ `notebook 6.X` available. There _is_ a point in having `notebook 7.X` next to `nbclassic`, so that could still be an addition (could be in this PR, but as mentioned before, I'm also fine just leaving that to someone else). As I understand from [here](https://jupyterhub.readthedocs.io/en/latest/howto/configuration/config-user-env.html) I should then probably set `c.Spawner.default_url` instead, if I want to make the interface provided by `nbclassic` (or `notebook 7`) the default.

Long story short: I think your PR is completely sane. If I figure out how to set the nbclassic interface as default, I'll report it here for future reference, but I think setting that `default_url` should do it.

**Comment by casparvl** [comment]
One more bread crumb for people coming from the traditional notebook interface: if they used `jupyter-nbextenstion list` before to list the enabled notebook extensions, that command is no longer there (since `notebook` is not installed - and I think it's `notebook` that normally installs this command). Neither is `jupyter-server-extension list`. However, one can use `jupyter-nbclassic-extension list` and `jupyter-nbclassic-serverextension list` for the same purpose.

**Review by casparvl — APPROVED** [review]
Lgtm! I'll submit some tests reports. Assuming those are succesful, this can be merged.

**Comment by casparvl** [comment]
@boegelbot please test @ jsc-zen2

**Comment by casparvl** [comment]
Test report by @casparvl
**SUCCESS**
Build succeeded for 5 out of 5 (5 easyconfigs in total)
tcn72.local.snellius.surf.nl - Linux RHEL 8.6, x86_64, AMD EPYC 7H12 64-Core Processor, Python 3.6.8
See https://gist.github.com/casparvl/4d5b0de1ab9b99cd2f4016c40ea8025d for a full test report.

**Comment by casparvl** [comment]
Going in, thanks @arielzn!

**Comment by lexming** [comment]
@casparvl Thanks for the heads up, LGTM
